src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,182 @@
+    /// </summary>
+    private readonly Dictionary<string, CustomMessageHandlerExtension> _extensions = new();
+
+    private readonly ICustomMessageHandlerFactory _customMessageHandlerFactory;
+
+    private readonly System.Runtime.Loader.AssemblyLoadContext _defaultLoadContext
+        = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(typeof(CustomMessageHandlerService).Assembly)
+        ?? throw new InvalidOperationException($"Cannot get assembly load context for {nameof(CustomMessageHandlerService)}.");
+
+    private readonly object _lockObject = new();
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CustomMessageHandlerService(ICustomMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public async ValueTask<string> HandleCustomMessageAsync(
+        Solution solution,
+        string assemblyFolderPath,
+        string assemblyFileName,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        _ = assemblyFolderPath ?? throw new ArgumentNullException(nameof(assemblyFolderPath));
+        _ = assemblyFileName ?? throw new ArgumentNullException(nameof(assemblyFileName));
+        _ = typeFullName ?? throw new ArgumentNullException(nameof(typeFullName));
+        _ = jsonMessage ?? throw new ArgumentNullException(nameof(jsonMessage));
+
+        var assemblyPath = Path.Combine(assemblyFolderPath, assemblyFileName);
+
+        Document? document = null;
+        if (documentId is not null)
+        {
+            document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(document);
+        }
+
+        try
+        {
+            ICustomMessageHandlerWrapper? handler;
+            lock (_lockObject)
+            {
+                // Check if the assembly is already loaded.
+                if (!_extensions.TryGetValue(assemblyFolderPath, out var extension))
+                {
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyFolderPath}", isCollectible: true);
+                    loadContext.Resolving += ResolveExtensionAssembly;
+
+                    extension = new CustomMessageHandlerExtension(loadContext);
+                    _extensions[assemblyFolderPath] = extension;
+                }
+
+                if (!extension.Handlers.TryGetValue(typeFullName, out handler))
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
```

Comment:
I'd avoid loading and executing code while holding the lock.

Reply:
Tagging @dibarbet on this since he gave the opposite feedback earlier.

Reply:
We'd still lock around the `_extensions` update (keep using Dictionary). We can make sure we load the assembly and call the constructor only when a new entry was added to the dictionary.

Reply:
I have split the lock, now the assembly loading and instance creation happens under a separate lock which will only affects other calls from the same extension.